src/crypto/index.js

```diff
@@ -1052,79 +1063,117 @@ Crypto.prototype._onRoomKeyRequestEvent = function(event) {
  *
  * @private
  */
-Crypto.prototype._processReceivedRoomKeyRequests = function() {
-    const requests = this._receivedRoomKeyRequests;
-    this._receivedRoomKeyRequests = [];
-    for (const req of requests) {
-        const userId = req.userId;
-        const deviceId = req.deviceId;
+Crypto.prototype._processReceivedRoomKeyRequests = async function() {
+    if (this._processingRoomKeyRequests) {
+        // we're still processing last time's requests; keep queuing new ones
+        // up for now.
+        return;
+    }
+    this._processingRoomKeyRequests = true;
 
-        const body = req.requestBody;
-        const roomId = body.room_id;
-        const alg = body.algorithm;
+    try {
+        // we need to grab and clear the queues in the synchronous bit of this method,
+        // so that we don't end up racing with the next /sync.
+        const requests = this._receivedRoomKeyRequests;
+        this._receivedRoomKeyRequests = [];
+        const cancellations = this._receivedRoomKeyRequestCancellations;
+        this._receivedRoomKeyRequestCancellations = [];
+
+        await Promise.map(
+            requests, (req) =>
+                this._processReceivedRoomKeyRequest(req),
+        );
+        await Promise.map(
```

Comment:
This is a beautiful clear explanation, thank you. Please add this as a comment to this code, as it is unclear why we're doing it in this order otherwise.